src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckShieldManager.scala

```diff
@@ -0,0 +1,42 @@
+package mesosphere.marathon
+package core.health.impl
+
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.state.Timestamp
+import mesosphere.marathon.core.health.HealthCheckShield
+
+import scala.concurrent.duration._
+import scala.collection.concurrent.TrieMap
+
+class HealthCheckShieldManager extends Object with StrictLogging {
+
+  private[this] val shields = TrieMap.empty[Task.Id, HealthCheckShield]
+
+  def isShielded(taskId: Task.Id): Boolean = {
```

Comment:
FAR: I think (i'm not 100% sure) most shield info will never get removed because marathon would call isShielded only for unhealthy instances. So for example, in most cases (where app becomes healthy again once heap dump is finished) the TLL is expired but it will never be removed.
Memory usage is quite low so it's acceptable, but there is an area of improvement here (like a quick shield check on HealthResult message) :)

Also, there are leader changes (= marathon leader restarts) quite often, so it should not be a big issue.

Reply:
Introduced periodical "purging" of the expired shields. 
It looks harder to react on tasks health events + introduces more coupling with existing marathon code.